File: frontend/src/types/schema.ts
export interface Space {
  id: number;
  name: string;
  vid_type: string;
}

export interface SpaceDetail {
  id: number;
  name: string;
  vid_type: string;
  partition_num: number;
  replica_factor: number;
  comment?: string;
  created_at: number;
  statistics: SpaceStatistics;
}

export interface SpaceStatistics {
  vertex_count?: number;
  edge_count?: number;
}

export interface PropertyDef {
  name: string;
  data_type: string;
  nullable: boolean;
  default_value?: string;
  comment?: string;
}

export interface Tag {
  id: number;
  name: string;
  properties: PropertyDef[];
  comment?: string;
  created_at: number;
}

export interface TagDetail {
  id: number;
  name: string;
  properties: PropertyDef[];
  indexes: IndexInfo[];
  created_at: number;
}

export interface EdgeType {
  id: number;
  name: string;
  properties: PropertyDef[];
  comment?: string;
  created_at: number;
}

export interface EdgeTypeDetail {
  id: number;
  name: string;
  properties: PropertyDef[];
  indexes: IndexInfo[];
  created_at: number;
}

export interface IndexInfo {
  id: number;
  name: string;
  index_type: string;
  entity_type: string;
  entity_name: string;
  fields: string[];
  comment?: string;
  created_at: number;
}

export interface CreateSpaceParams {
  name: string;
  vid_type?: string;
  comment?: string;
}

export interface CreateTagParams {
  name: string;
  properties: PropertyDef[];
  ttlCol?: string;
  ttlDuration?: number;
}

export interface CreateEdgeTypeParams {
  name: string;
  properties: PropertyDef[];
  ttlCol?: string;
  ttlDuration?: number;
}

export interface CreateIndexParams {
  name: string;
  index_type: string;
  entity_type: string;
  entity_name: string;
  fields: string[];
  comment?: string;
}

// Extended types for Phase 4 & 5

export type DataType =
  | 'STRING'
  | 'INT64'
  | 'DOUBLE'
  | 'BOOL'
  | 'DATETIME'
  | 'DATE'
  | 'TIME'
  | 'TIMESTAMP';

export interface Property {
  name: string;
  type: DataType;
  default_value?: string;
  nullable?: boolean;
}

export type IndexStatus = 'creating' | 'finished' | 'failed' | 'rebuilding';

export interface Index {
  id: number;
  name: string;
  type: 'TAG' | 'EDGE';
  schemaName: string;
  properties: string[];
  status: IndexStatus;
  created_at: number;
  updated_at?: number;
  progress?: number;
  errorMessage?: string;
}

export interface IndexStats {
  total: number;
  byType: {
    tag: number;
    edge: number;
  };
  byStatus: {
    creating: number;
    finished: number;
    failed: number;
    rebuilding: number;
  };
}

export interface UpdateTagParams {
  add_properties?: PropertyDef[];
  drop_properties?: string[];
}

export interface UpdateEdgeTypeParams {
  add_properties?: PropertyDef[];
  drop_properties?: string[];
}

export interface DDLData {
  space: string;
  tags: string[];
  edges: string[];
  indexes: string[];
}

File: crates/graphdb-api/src/api/server/web/models/schema.rs
//! Schema Extension Models

use serde::{Deserialize, Serialize};

/// Space detail response
#[derive(Debug, Serialize)]
pub struct SpaceDetail {
    pub id: u64,
    pub name: String,
    pub vid_type: String,
    pub partition_num: i32,
    pub replica_factor: i32,
    pub comment: Option<String>,
    pub created_at: i64,
    pub statistics: SpaceStatistics,
}

/// Space statistics
#[derive(Debug, Serialize)]
pub struct SpaceStatistics {
    pub tag_count: i64,
    pub edge_type_count: i64,
    pub index_count: i64,
    pub estimated_vertex_count: i64,
    pub estimated_edge_count: i64,
}

/// Tag summary
#[derive(Debug, Serialize)]
pub struct TagSummary {
    pub id: i64,
    pub name: String,
    pub property_count: i64,
    pub index_count: i64,
    pub created_at: i64,
}

/// Tag detail
#[derive(Debug, Serialize)]
pub struct TagDetail {
    pub id: i64,
    pub name: String,
    pub properties: Vec<PropertyDef>,
    pub indexes: Vec<IndexInfo>,
    pub created_at: i64,
}

/// Edge type summary
#[derive(Debug, Serialize)]
pub struct EdgeTypeSummary {
    pub id: i64,
    pub name: String,
    pub property_count: i64,
    pub index_count: i64,
    pub created_at: i64,
}

/// Edge type detail
#[derive(Debug, Serialize)]
pub struct EdgeTypeDetail {
    pub id: i64,
    pub name: String,
    pub properties: Vec<PropertyDef>,
    pub indexes: Vec<IndexInfo>,
    pub created_at: i64,
}

/// Property definition
#[derive(Debug, Serialize, Deserialize)]
pub struct PropertyDef {
    pub name: String,
    pub data_type: String,
    pub nullable: bool,
    pub default_value: Option<String>,
}

/// Index information
#[derive(Debug, Serialize)]
pub struct IndexInfo {
    pub id: i64,
    pub name: String,
    pub index_type: String, // INDEX, UNIQUE, FULLTEXT
    pub fields: Vec<String>,
    pub status: String, // ACTIVE, BUILDING, FAILED
    pub progress: Option<i32>,
    pub created_at: i64,
}

/// Create index request
#[derive(Debug, Deserialize)]
pub struct CreateIndexRequest {
    pub name: String,
    pub index_type: String,
    pub entity_type: String, // TAG or EDGE
    pub entity_name: String,
    pub fields: Vec<String>,
    pub comment: Option<String>,
}

/// Update tag request
#[derive(Debug, Deserialize)]
pub struct UpdateTagRequest {
    pub add_properties: Option<Vec<PropertyDef>>,
    pub drop_properties: Option<Vec<String>>,
}

/// Update edge type request
#[derive(Debug, Deserialize)]
pub struct UpdateEdgeTypeRequest {
    pub add_properties: Option<Vec<PropertyDef>>,
    pub drop_properties: Option<Vec<String>>,
}
